src/code/RepositorySettings.cs

```
@@ -168,28 +168,27 @@ public static PSRepositoryInfo Update(string repoName, Uri repoUri, int repoPrio
 
                 // determine if existing repository node (which we wish to update) had Url to Uri attribute
                 Uri thisUrl = null;
-                if (repoUri != null)
-                {
-                    if (urlAttributeExists)
-                    {
+                if (urlAttributeExists) {
+                    if (repoUri != null) {
                         node.Attribute("Url").Value = repoUri.AbsoluteUri;
-                        // Create Uri from node Uri attribute to create PSRepositoryInfo item to return.
-                        if (!Uri.TryCreate(node.Attribute("Url").Value, UriKind.Absolute, out thisUrl))
-                        {
-                            throw new PSInvalidOperationException(String.Format("Unable to read incorrectly formatted Url for repo {0}", repoName));
-                        }
                     }
-                    else
+                    if (!Uri.TryCreate(node.Attribute("Url").Value, UriKind.Absolute, out thisUrl))
```

Your code sets the node's Uri (or Url) attribute to the Uri passed in by the user, then we try to validate the node's new Uri. If it fails we throw but we've changed the node.Attribute(Uri) (or Url) already and don't revert it back. On the plus side though, we *always* validate that the value present for Uri (or Url) is valid, which the original code didn't do.

Looping @alerickson @PaulHigin into this for what they also think the flow should be, but I think if we detect the new Uri value provided by the user isn't a valid Uri we should throw an error and keep the original. If the user isn't trying to set the Uri, in my opinion, we should still validate the existing Uri (or Url) and throw if needed, but keep the original value.

---

One way to do what I suggested could be that in the original code, we had a condition for `if (repoUri != null)`...I think we need an `else` to that condition which would try to validate the current Uri- throw if invalid and if valid then set `thisUri` to the current Uri value. This way when we create `updatedRepo` and use `thisUri` it's not possible for `thisUri` to be null.

---

I mentioned this a little earlier, but I don't think we need to validate what the current contents of uri/url is.  I think you can just put lines 175-178 before lines 172-174 so that the error gets thrown before the property gets set. Same with 184-187 getting placed before 181-183.  That should fix the  original problem right?  Good catch on this bug though

---

I feel the creation of the Uri from user input should be done before checking whether an existing Uri exists in the store (Can we ever have a repository store entry without a Uri?  If not then this check should not be needed).  I see that the `Uri.TryCreate()` is called twice, but this can be done just once and throw if needed, before anything else is done (since that is a terminating error condition).

---

> Can we ever have a repository store entry without a Uri? 
Nope, only if a user manually modifies their file

---

The reason why `Uri.TryCreate()` is called twice in my previous code is because there are 2 different conditions, one for Uri and the other for Url. I updated the code and now `Uri.TryCreate()` is called 4 times because for each condition, I need to check if the user input Uri/Url is a valid one before I update the value. I call `Uri.TryCreate()` before `if (repo != null)` because if user doesn't input a Uri, `thisUri` will be `null`, which causes the current bug.

---

I feel need to only call `Uri.TryCreate()` one time for the Uri provided by the user.  Also I think we only need to update the existing Uri or Url if the user provided one (see my comment below), and do nothing otherwise.